src/DataLib/fendl32B_retrofit/GROUPR/groupr_tools.py

```diff
@@ -0,0 +1,180 @@
+    # Read in ENDF tape using ENDFtk
+    tape = ENDFtk.tree.Tape.from_file(endf_path)
+
+    # Determine the material ID
+    mat_ids = tape.material_numbers
+    matb = mat_ids[0]
+
+    # Set MF for cross sections
+    xs_MF = 3
+
+    # Extract out the file
+    file = tape.material(matb).file(xs_MF).parse()
+
+    # Extract the MT numbers that are present in the file
+    MTs = [MT.MT for MT in file.sections.to_list()]
+
+    return matb, MTs
+
+# Define a function to format GROUPR input cards
+def format_card(card_name, card_content, MTs):
+    card_str = ''
+    gen_str = ' ' + ' '.join(map(str, card_content))
+    if card_name == 9:
+        card_str = ' ' + '/\n '.join(card_content) + '/\n'
+    elif card_name == 4:
+        card_str += gen_str + '\n'
+    else:
+        card_str += gen_str + '/\n'
+    return card_str
+
+# Define a function to create the GROUPR input file
+def groupr_input_file_format(matb, MTs, element, A, mt_table):
+
+    cards = {}
+
+    # Set Card 1
+    nendf = 20 # unit for endf tape
+    npend = 21 # unit for pendf tape
+    ngout1 = 0 # unit for input gout tape (default=0)
+    ngout2 = 31 # unit for output gout tape (default=0)
+
+    cards[1] = [nendf, npend, ngout1, ngout2]
+
+    # Set Card 2
+    # matb -- (already defined) -- material to be processed
+    ign = 17 # neutron group structure option
+    igg = 0 # gamma group structure option
+    iwt = 11 # weight function option
+    lord = 0 # Legendgre order
+    ntemp = 1 # number of temperatures (default = 1)
+    nsigz = 1 # number of sigma zeroes (default = 1)
+    iprint = 1 # long print option (0/1=minimum/maximum) -- (default=1)
+    ismooth = 1 # swith on/off smoother operation (1/0, default=1=on)
+
+    cards[2] = [matb, ign, igg, iwt, lord, ntemp, nsigz, iprint]
+
+    # Set Card 3
+    Z = str(elements.index(element) + 1).zfill(2)
+    title = f'"{Z}-{element}-{A} for TENDL 2017"'
+    cards[3] = [title]
+
+    # Set Card 4
+    temp = 293.16 # temperature in Kelvin
+    cards[4] = [temp]
+
+    # Set Card 5
+    sigz = 0 # sigma zero values (including infinity)
+    cards[5] = [sigz]
+
+    # Set Card 9
+    mfd = 3 # file to be processed
+    mtd = MTs # sections to be processed
+    cards[9] = []
+    for mt in MTs:
+        mtname = mt_table[mt_table['MT'] == mt]['Reaction'].values[0] # description of section to be processed
+        card9_line = f'{mfd} {mt} "{mtname}"'
+        cards[9].append(card9_line)
+
+    # Set Card 10
+    matd = 0 # next mat number to be processed
+    cards[10] = [matd]
+
+    return cards
+
+# Define a function to write out the GROUPR input file
+def groupr_input_file_writer(cards, MTs):
+    # Write the input deck to the groupr.inp file
+    with open('groupr.inp', 'w') as f:
+        f.write('groupr\n')
+        max_card_index = 10
+        for i in range(max_card_index + 1):
+            try:
+                f.write(format_card(i, cards[i], MTs))
+            except KeyError:
+                continue
```

Comment:
I think you can take advantage of the dictionary this way...

```suggestion
        for card_num, card in cards.items():
            f.write(format_card(i, cards[i], MTs))
```